### PR TITLE
wasm: Support debug builds

### PR DESCRIPTION
**Pull request by patrick-east**
We now have a flag for building with lower optimizations enabled and
debug symbols. In addition there are some changes to key off a DEBUG
build flag in the source controlling a trace helper and heap checks.

README is updated to explain the usage.

Signed-off-by: Patrick East <east.patrick@gmail.com>

### CODE REVIEW

**Comment by patrick-east** [comment]
I snuck in a change for the node version with this too

**Review by tsandall — DISMISSED** [review]
LGTM. I haven't tried it out all the way but the changes look fine. Were you able to test and see if the `opa build` leaves the debug symbols intact?

File: wasm/src/std.h [hunk 1]
```diff
@@ -10,4 +10,16 @@
 void opa_abort(const char *msg);
 void opa_println(const char *msg);
 
+#ifdef DEBUG
+#include "printf.h"
+#define TRACE(...)                               \
```
Comment (koponen-styra) [inline wasm/src/std.h:15]: I'd wrap a multi line macro contents in do - while as follows:

#define TRACE(...) \
   do {
      char __trace_buf ...
   } while (0);

This guarantees you can use the macro anywhere.

Comment (patrick-east) [reply wasm/src/std.h:15]: Thats good to know 👍   Will update it.

**Comment by patrick-east** [comment]
> Were you able to test and see if the opa build leaves the debug symbols intact?

Not yet, will open an issue to track supporting it if we are losing them.

File: wasm/src/std.h [hunk 2]
```diff
@@ -10,4 +10,18 @@
 void opa_abort(const char *msg);
 void opa_println(const char *msg);
 
+#ifdef DEBUG
+#include "printf.h"
+#define TRACE(...)                                   \
+    {                                                \
```
Comment (koponen-styra) [inline wasm/src/std.h:16]: Sorry, I wasn't clear enough. This looks OK but you should remove the outer curly brackets and then also remove the semi-colon after "while". Then you can use "TRACE();" in any context, like any statement.

Comment (patrick-east) [reply wasm/src/std.h]: I figured there was some trick with the do-while inside the block, but using it as the wrapper makes way more sense 😆

**Review by koponen-styra — APPROVED** [review]

**Pull request merged**